site/recipes/recipe_different_processing_against_stream.md

```diff
@@ -0,0 +1,150 @@
+
+In this instance, we can apply different processing against the stream of mileage readings from the sensor and generate warnings when poor gas mileage is detected.
+
+## Setting up the application
+
+We assume that the environment has been set up following the steps outlined in the [Getting Started Guide](../docs/quarks-getting-started). Let's begin by creating a `DirectProvider` and `Topology`. We choose a `DevelopmentProvider` so that we can view the topology graph using the console URL (refer to the [Application Console](../docs/console) page for a more detailed explanation of this provider). The initial mileage value has also been defined.
+
+```java
+    public class TruckSensor {
+        /**
+        * Hypothetical value for initial gas mileage
+        */
+        static double currentMileage = 10.5;
+
+        public static void main(String[] args) throws Exception {
+
+            DirectProvider dp = new DevelopmentProvider();
+
+            System.out.println(dp.getServices().getService(HttpServer.class).getConsoleUrl());
+
+            Topology top = dp.newTopology("TruckSensor");
+
+            // The rest of the code pieces belong here.
+        }
+    }
+```
+
+## Generating mileage sensor readings
+
+The next step is to simulate a stream of gas mileage readings. In our `main()`, we use the `poll()` method to generate a flow of tuples (readings), where each tuple arrives every second.
+
+```java
+    // Generate a stream of mileage sensor readings.
+    Random r = new Random();
+    TStream<Double> mileage = top.poll(() -> {
+        // Change current mileage by some random amount between -0.4 and 0.4.
+        double newMileage = -0.4 + (0.4 + 0.4) * r.nextDouble() + currentMileage;
+        currentMileage = newMileage;
+        return currentMileage;
+    }, 1, TimeUnit.SECONDS);
+```
+
+## Applying different processing to the stream
+
+The company can now perform analytics on the `mileage` stream to generate poor gas mileage warnings by starting with the original stream and altering it. First, we filter out values that are out of range. Then, we tag the stream with the _mileage_ tag for easier viewing in the console (we add this tag after every step). Next, we truncate the reading to one decimal place. We then use `peek` to print out the current mileage so that we can get a sense of the truck's current state. After, we keep the mileage readings that are considered "poor." Finally, we terminate the stream by printing out warnings for poor gas mileage.
+
+```java
+    DecimalFormat df = new DecimalFormat("#.#");
+
+    // Perform analytics on mileage readings to generate poor gas mileage warnings.
+    mileage = mileage.filter(tuple -> tuple >= 7.0 && tuple <= 14.0);
```

Comment:
Seems like a good opportunity to demonstrate, and promote, the use of the fluent style here?  It would remove lots of visual noise from all of the unnecessary `mileage = mileage. ...` . e.g., 
```
mileage.filter(...).tag("mileage")
       .modify(...).tag("mileage")
       .peek(...).tag("mileage")
       .filter(...).tag("mileage")
       .sink(...);
```